poetry/installation/authenticator.py

```diff
@@ -0,0 +1,133 @@
+from typing import TYPE_CHECKING
+
+from poetry.utils._compat import urlparse
+from poetry.utils.password_manager import PasswordManager
+
+
+if TYPE_CHECKING:
+    from typing import Any
+    from typing import Optional
+    from typing import Tuple
+
+    from clikit.api.io import IO
+    from requests import Request  # noqa
+    from requests import Response  # noqa
+    from requests import Session  # noqa
+
+    from poetry.config.config import Config
+
+
+class Authenticator(object):
+    def __init__(self, config, io):  # type: (Config, IO) -> None
+        self._config = config
+        self._io = io
+        self._session = None
+        self._credentials = {}
+        self._password_manager = PasswordManager(self._config)
+
+    @property
+    def session(self):  # type: () -> Session
+        from requests import Session  # noqa
+
+        if self._session is None:
+            self._session = Session()
+
+        return self._session
+
+    def request(self, method, url, **kwargs):  # type: (str, str, Any) -> Response
+        from requests import Request  # noqa
+        from requests.auth import HTTPBasicAuth
+
+        request = Request(method, url)
+
+        username, password = self._get_credentials_for_url(url)
+
+        if username is not None and password is not None:
+            request = HTTPBasicAuth(username, password)(request)
+
+        session = self.session
+        prepared_request = session.prepare_request(request)
+
+        proxies = kwargs.get("proxies", {})
+        stream = kwargs.get("stream")
+        verify = kwargs.get("verify")
+        cert = kwargs.get("cert")
+
+        settings = session.merge_environment_settings(
+            prepared_request.url, proxies, stream, verify, cert
+        )
+
+        # Send the request.
+        send_kwargs = {
+            "timeout": kwargs.get("timeout"),
+            "allow_redirects": kwargs.get("allow_redirects", True),
+        }
+        send_kwargs.update(settings)
+        resp = session.send(prepared_request, **send_kwargs)
+
+        resp.raise_for_status()
+
+        return resp
+
+    def _get_credentials_for_url(
```

Comment:
Should the following methods move to a more general use location like utils? Might not be required if we intend to only use creds here. We can also defer this decision.

Reply:
I don't foresee this to be used elsewhere at the moment. If that were to be the case we can always refactor later.